src/main/scala/is/hail/utils/richUtils/RichIndexedRowMatrix.scala

```diff
@@ -0,0 +1,160 @@
+package is.hail.utils.richUtils
+
+import org.apache.spark.Partitioner
+import org.apache.spark.mllib.linalg.{DenseMatrix, Matrix}
+import org.apache.spark.mllib.linalg.distributed.{BlockMatrix, IndexedRowMatrix}
+import org.apache.spark.rdd.RDD
+
+/**
+  * Adds a better toBlockMatrix method to IndexedRowMatrix
+  */
+
+object RichIndexedRowMatrix {
+
+  implicit class RichIndexedRowMatrix(indexedRowMatrix: IndexedRowMatrix) {
+    def toBlockMatrixDense(): BlockMatrix = {
+      toBlockMatrixDense(1024, 1024)
+    }
+
+    def toBlockMatrixDense(rowsPerBlock: Int, colsPerBlock: Int): BlockMatrix = {
+      require(rowsPerBlock > 0,
+        s"rowsPerBlock needs to be greater than 0. rowsPerBlock: $rowsPerBlock")
+      require(colsPerBlock > 0,
+        s"colsPerBlock needs to be greater than 0. colsPerBlock: $colsPerBlock")
+      val m = indexedRowMatrix.numRows().toInt
+      val n = indexedRowMatrix.numCols().toInt
+
+      val blocksRDD: RDD[((Int, Int), Matrix)] = indexedRowMatrix.rows.flatMap({ ir =>
+        val blockRow = ir.index / rowsPerBlock
+        val rowInBlock = ir.index % rowsPerBlock
+
+        val partitionedArray: Iterator[Array[Double]] = ir.vector.toArray.grouped(colsPerBlock)
+        val partitionedArrayWithIndices = partitionedArray.zipWithIndex
+
+        partitionedArrayWithIndices.map(tuple =>
+          tuple match {
+            case (values, blockColumn) =>
+              ((blockRow.toInt, blockColumn), (rowInBlock.toInt, values))
+          })
+      }).groupByKey(new GridPartitioner(m, n, rowsPerBlock, colsPerBlock)).map(tuple =>
+        tuple match {
+          case ((blockRow, blockColumn), itr) =>
+            val actualNumRows =
+              if (m - blockRow * rowsPerBlock < rowsPerBlock) {
+                m - blockRow * rowsPerBlock
+              } else {
+                rowsPerBlock
+              }
+            val actualNumColumns =
+              if (n - blockColumn * colsPerBlock < colsPerBlock) {
+                n - blockColumn * colsPerBlock
+              } else {
+                colsPerBlock
+              }
```

Comment:
I feel like we should precompute these outside the lambda:

```scala
val m = indexedRowMatrix.numRows().toInt
val n = indexedRowMatrix.numCols().toInt
val lastRowBlockIndex = m / rowsPerBlock
val lastColBlockIndex = n / colsPerBlock
val lastRowBlockSize = m % rowsPerBlock
val lastColBlockSize = n % colsPerBlock
```

and then have a simple `if` here that switches on `blockRow == lastRowBlockIndex`